ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java

```diff
@@ -101,42 +109,55 @@
   CosmosDataAccessor(CloudConfig cloudConfig, AzureCloudConfig azureCloudConfig, VcrMetrics vcrMetrics,
       AzureMetrics azureMetrics) {
     // Set up CosmosDB connection, including retry options and any proxy setting
-    ConnectionPolicy connectionPolicy = new ConnectionPolicy();
-    // TODO: would like to use different timeouts for queries and single-doc reads/writes
-    connectionPolicy.setRequestTimeoutInMillis(cloudConfig.cloudQueryRequestTimeout);
     // Note: retry decisions are made at CloudBlobStore level.  Configure Cosmos with no retries.
-    RetryOptions noRetries = new RetryOptions();
-    noRetries.setMaxRetryAttemptsOnThrottledRequests(0);
-    connectionPolicy.setRetryOptions(noRetries);
+    ThrottlingRetryOptions throttlingRetryOptions = new ThrottlingRetryOptions();
+    throttlingRetryOptions.setMaxRetryAttemptsOnThrottledRequests(0);
+    CosmosClientBuilder cosmosClientBuilder = new CosmosClientBuilder().endpoint(azureCloudConfig.cosmosEndpoint)
+        .key(getCosmosKey(azureCloudConfig))
+        .throttlingRetryOptions(throttlingRetryOptions)
+        .consistencyLevel(com.azure.cosmos.ConsistencyLevel.SESSION);
+
+    // It looks like we can no longer configure request timeouts on GateWayConnection mode
+    // since cosmos team has hidden them in latest versions. (https://github.com/Azure/azure-sdk-for-java/pull/11702).
+    // Default value seems to be 60 seconds. For DirectConnection (which uses TCP to talk directly to cosmos storage nodes),
+    // the value can be configured but has lower limit of 5 seconds and upper limit of 10 seconds.
+    // We can probably reach out to them if we see requests taking longer than these defaults.
+    // For reference, here is difference between Gateway and direct connection modes -
+    // https://docs.microsoft.com/en-us/azure/cosmos-db/sql/sql-sdk-connection-modes
+    GatewayConnectionConfig gatewayConnectionConfig = new GatewayConnectionConfig();
+    if (cloudConfig.vcrProxyHost != null) {
+      gatewayConnectionConfig.setProxy(new ProxyOptions(ProxyOptions.Type.HTTP,
+          new InetSocketAddress(cloudConfig.vcrProxyHost, cloudConfig.vcrProxyPort)));
+    }
     if (azureCloudConfig.cosmosDirectHttps) {
       logger.info("Using CosmosDB DirectHttps connection mode");
-      connectionPolicy.setConnectionMode(ConnectionMode.Direct);
-    }
-    if (cloudConfig.vcrProxyHost != null) {
-      connectionPolicy.setProxy(cloudConfig.vcrProxyHost, cloudConfig.vcrProxyPort);
+      cosmosClientBuilder.directMode(new DirectConnectionConfig(), gatewayConnectionConfig);
+    } else {
+      cosmosClientBuilder.gatewayMode(gatewayConnectionConfig);
```

Comment:
Nit: maybe initialize the GatewayConnectionConfig here.

Reply:
It looks like GatewayConnectionConfig is also being used in their direct connection mode (for some of the metadata operations). Due to that, I had to initialize it outside the `if` condition.